text/0000-update-version-handling.md

```diff
@@ -0,0 +1,225 @@
+impl ToolSpec {
+  fn try_from_str(string: &str) -> Result<ToolSpec, String> {
+    // split apart the tool name and its version
+  }
+}
+```
+
+This is a relatively simple implementation. The only complication is the existence of Node namespaces: `@namespace/package`. This means parsing the version is not as simple as splitting on `"@"`. However, it should still be straightforward using the combination of a regular expression and the existing semver tooling.
+
+The `ToolSpec` change will make the format `tool[@version]` work; updating the commands to take a `Vec<ToolSpec>` will support passing multiple arguments.
+
+# Critique
+[critique]: #critique
+
+This introduces breaking changes—something which should not be done lightly! It does so under the assumption that fetching or especially installing or pinning multiple tools simultaneously is a relatively common operation. This assumption may be wrong. It may also be the case that implementing this—with the required changes to 
+
+It also introduces a move away from users’ existing habits with `nvm` and `nodenv` and similar tools. This is thus an increase in the learning/switching costs for the tool.
+
+*Not* doing this is not likely to be a blocker for wider adoption. The “problem” identified here represents a papercut, not a major pain point.
+
+## Alternatives
+
+-   Support *both* invocation styles. This may *just* be technically feasible, allowing users to pass any of the following invocations successfully:
+
+    ```sh
+    notion pin create-react-app
+    notion pin create-react-app@3.0
+    notion pin create-react-app 3.0
+    notion pin create-react-app typescript
+    notion pin create-react-app@3.0 typescript
+    notion pin create-react-app typescript@3.4
+    notion pin create-react-app@3.0 typescript@3.4
+    ```
+
+    However, this introduces substantial additional complexity for the implementation, and ambiguity around certain invocations, e.g. for tagged versions/special names. For example, how should the following invocation be resolved?
+    
+    ```sh
+    notion pin create-react-app latest
+    ```
+
+    Does it refer to the latest version of `create-react-app`, or the default version of `create-react-app` and [the `latest` package](https://www.npmjs.com/package/latest)? For all non-SemVer-string parses, this will be impossible to resolve.
+
+    The proposed design here is well-proven with other tools which install more than simply a single tool-and-version pair, e.g. `npm`, `yarn`, and `cargo-edit`, and it does not fall prey to these problems of indeterminacy.
+
+-   Leave things as they are. This has the upside of matching other Node version managers, all of which use a close variant on our current syntax.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- This will *suggest* to users that they can use tags, e.g. `tool@next`, a feature we do not currently support. Should we include support for that here, or plan it as follow-on work, or neither?
```

Comment:
I think this should be planned as separate work. This is something we could already implement separate from the changes here, and this RFC would get the benefit of that. (Assuming the changes in this RFC will re-use most of the existing code to parse the input version from the command line)

Reply:
That's my take as well, but I wanted to raise it. I also learned in poking through the implementation that we *do* have `lts` support, so I'll make sure the reference to that in the table is updated!

Reply:
Agreed. We'll also need to figure out how to handle tags on the `node` version, which doesn't have the same npm registry concepts that other tools have, since it's a level above `npm`.

Reply:
Agreed. We'll also need to figure out how to handle tags on the node version, which doesn't have the same npm registry concepts that other tools have, since it's a level above `npm`.

Reply:
@charlespierce yeah, I think that warrants a further round of design. My inclination is actually to break apart the notion of versions there a bit: *package* versions are not identical with *platform* versions, though given how npm and Yarn work they overlap in complicated ways.

Reply:
Moved this to a <b>Future Work</b> subsection (e183bf13).